cmd/oci-trace-hook/trace.go

```diff
@@ -0,0 +1,276 @@
+		Dir: ".",
+		Env: os.Environ(),
+		Files: []*os.File{
+			os.Stdin,
+			nil,
+			nil,
+		},
+		// Sys: sysproc,
+	}
+	if pid > 0 {
+		process, err := os.StartProcess("/usr/libexec/oci/hooks.d/oci-trace-hook", []string{"/usr/libexec/oci/hooks.d/trace", "-r", strconv.Itoa(pid), "-f", fileName}, attr)
+		if err != nil {
+			return fmt.Errorf("cannot launch process err: %q", err.Error())
+		}
+		time.Sleep(2 * time.Second) // Waits 2 seconds to compile using clang and llvm
+		processPID := process.Pid
+		f, err := os.Create("pid")
+		if err != nil {
+			return fmt.Errorf("cannot write pid to file err:%q", err.Error())
+		}
+		defer f.Close()
+		f.WriteString(strconv.Itoa(processPID))
+		err = process.Release()
+		if err != nil {
+			return fmt.Errorf("cannot detach process err:%q", err.Error())
+		}
+	} else {
+		return fmt.Errorf("container not running")
+	}
+	return nil
+}
+
+// run the BPF source and attach it to raw_syscalls:sys_enter tracepoint
+func runBPFSource(pid int, fileName string) error {
+	logrus.Println("Running floating process PID to attach:", pid)
+	syscalls := make(calls, 303)
+	src := strings.Replace(source, "$PARENT_PID", strconv.Itoa(pid), -1)
+	m := bpf.NewModule(src, []string{})
+	defer m.Close()
+
+	tracepoint, err := m.LoadTracepoint("enter_trace")
+	if err != nil {
+		return err
+	}
+
+	logrus.Println("Loaded tracepoint")
+
+	if err := m.AttachTracepoint("raw_syscalls:sys_enter", tracepoint); err != nil {
+		return fmt.Errorf("unable to load tracepoint err:%q", err.Error())
+	}
+
+	table := bpf.NewTable(m.TableId("events"), m)
+	channel := make(chan []byte)
+	perfMap, err := bpf.InitPerfMap(table, channel)
+	if err != nil {
+		return fmt.Errorf("unable to init perf map err:%q", err.Error())
+	}
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt)
+	rsc := false // Reached Seccomp syscall
+	go func() {
+		var e event
+		for {
+			data := <-channel
+			err := binary.Read(bytes.NewBuffer(data), binary.LittleEndian, &e)
+			if err != nil {
+				logrus.Errorf("failed to decode received data '%s': %s\n", data, err)
+				continue
+			}
+			name, err := getName(e.ID)
+			if err != nil {
+				logrus.Errorf("failed to get name of syscall from id : %d received : %q", e.ID, name)
+			}
+			if name == "seccomp" {
+				rsc = true
+				continue
+			}
+			if rsc {
+				syscalls[name]++
+			}
+		}
+	}()
+	logrus.Println("PerfMap Start")
+	perfMap.Start()
+	<-sig
+	logrus.Println("PerfMap Stop")
+	perfMap.Stop()
+	if err := generateProfile(syscalls, fileName); err != nil {
+		return err
+	}
+	return nil
+}
+
+// send SIGINT to the floating process reading the perfbuffer
+func sendSIGINT() error {
+	f, err := ioutil.ReadFile("pid")
+
+	if err != nil {
+		return err
+	}
+
+	Spid := string(f)
+
+	pid, _ := strconv.Atoi(Spid)
+	p, _ := os.FindProcess(pid)
+	p.Signal(os.Interrupt)
+	return nil
+}
+
+// generate the seccomp profile from the syscalls provided
+func generateProfile(c calls, fileName string) error {
```

Comment:
We're currently also dumping the syscalls from runc. What we can do to filter them is only dump syscalls that come after the first `execve`.

Reply:
This is still valid. What we need to do is to find the index of the _first_ `execve` and only consider syscalls _after_ that.

To increase confidence, we could add unit tests (i.e., a `cmd/oci-trace-hook/trace_test.go`) but that has a low priority atm.

Reply:
I tried adding the calls after the first `execve`, but the profile didn't work.

Reply:
The problem is that the events are not sent in order, so we might need to move the logic into BPF.

Reply:
```
# First run
[libpod (pull-request-3576)]$ cat prof.json 
{"defaultAction":"SCMP_ACT_ERRNO","syscalls":[{"names":["0_fcntl","1_exit_group","2_writev","3_getdents64","4_ope
n","5_nanosleep","6_getuid","7_set_tid_address","8_epoll_pwait","9_lstat","10_execve","11_mprotect","12_stat","13
_close","14_arch_prctl","15_ioctl"],"action":"SCMP_ACT_ALLOW","args":[],"comment":"","includes":{},"excludes":{}}
]}
# Second run
[libpod (pull-request-3576)]$ cat prof.json 
{"defaultAction":"SCMP_ACT_ERRNO","syscalls":[{"names":["0_arch_prctl","1_stat","2_ioctl","3_lstat","4_open","5_m
protect","6_nanosleep","7_set_tid_address","8_getdents64","9_exit_group","10_getuid","11_fcntl","12_close","13_ex
ecve","14_writev"],"action":"SCMP_ACT_ALLOW","args":[],"comment":"","includes":{},"excludes":{}}]}[libpod (pull-r
equest-3576)]$ 
```

Reply:
Ah. Sorry, the problem should be that `syscalls` is a map, so we loose the order.

Reply:
I did some more digging. Preserving the order and recording _all_ syscalls, we get the following slice:
```json
"names":["0_nanosleep","1_futex","2_brk","3_epoll_pwait","4_openat","5_epoll_ctl","6_getdents64","7_newfstatat","8_close","9_fcntl","10_capget","11_read","12_prctl","13_fstat","14_fchown","15_setgroups","16_setgid","17_setuid","18_capset","19_chdir","20_getppid","21_write","22_execve","23_arch_prctl","24_set_tid_address","25_mprotect","26_getuid","27_ioctl","28_stat","29_open","30_lstat","31_writev","32_exit_group"]                                                                                    
```

On my machine, the seccomp profile starts working starting with syscall **15** (i.e. `setgroups`).

All previous syscalls are already in the default seccomp profile, so we are not getting less secure (or more secure) preserving them. @giuseppe @rhatdan, how stable do you think it is to assume that starting with `setgroups` runc will keep working (also in the future). I am tempted to just record all the syscalls to remain forward compatible.

Reply:
I think we should start recording after `prctl(PR_SET_SECCOMP)`.  That is where the seccomp profile is installed so anything afterwards must be recorded and allowed